hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergedReadHandle.java

```diff
@@ -166,6 +167,17 @@ private List<HoodieRecord<T>> doMergedRead(Option<HoodieFileReader> baseFileRead
         String key = record.getRecordKey();
         if (deltaRecordMap.containsKey(key)) {
           deltaRecordKeys.remove(key);
+
+          // When we find a processing time based delete, e.g., a delete record without valid ordering value,
+          // or with default ordering value, i.e., 0, we can safely ignore records before the delete record.
+          // That means, we can return the log record without merging with the record from base file.
+          if (deltaRecordMap.get(key).getMetaDataInfo(DELETE_FOUND_WITHOUT_ORDERING_VALUE).isPresent()) {
+            if (!deltaRecordMap.get(key).isDelete(readerSchema, config.getProps())) {
```

Comment:
I assume only deletes enter here so there is no need to check whether the delta record is a delete or not.

Reply:
data record can also be here, since this information need to be kept until the merging with the record from base file. After the delete, new insert or update records could be added for the same key. In more details, suppose we have a base file, and 4 log files. for rk, we have records, bfr, lfr1, lfr2, lfr3(the delete), lfr4. Since in the end bfr should be merged with lfr4, lfr4 should have this meta info even lfr4 is an insert or update.